package.json

```diff
@@ -3,7 +3,11 @@
     "version": "7.0.0",
     "description": "OpenFin Workspace Starter",
     "scripts": {
-        "package-how-tos": "node ./scripts/package.js"
+        "package-how-tos": "node ./scripts/package.js legacy",
+        "build": "node ./scripts/package.js build",
```

Comment:
@obany has come up with a suggestion of using npm workspaces to perform the build. This should speed up the npm install time of a build.

Reply:
I'm not really familiar with the `npm` ecosystem. I will read up on `npm workspaces` before making the change. Stay tuned.

Reply:
Seems `workspaces` are available since `npm v7.x`. We're using the `node:14` image which has `npm v6.x`.

I tried on `node:16` with `npm v8x` and got several errors when building HOWTO packages:
```
ERROR in unable to locate '/root/work/how-to/integrate-with-salesforce/node_modules/@openfin/salesforce-lwc/preload.js' glob
```
```
ERROR in /root/work/how-to/support-context-and-intents/client/src/platform.ts
./client/src/platform.ts 23:8-31
[tsl] ERROR in /root/work/how-to/support-context-and-intents/client/src/platform.ts(23,9)
      TS2322: Type '(InteropBroker: Constructor<InteropBroker>, provider: any, options: any, ...args: any[]) => Promise<PlatformInteropBroker>' is not assignable to type 'OverrideCallback<InteropBroker, InteropBroker>'.
```
```
ERROR in /root/work/how-to/customize-workspace/client/src/bootstrapper.ts
./client/src/bootstrapper.ts 45:8-18
[tsl] ERROR in /root/work/how-to/customize-workspace/client/src/bootstrapper.ts(45,9)
      TS2322: Type '(view: string | OpenFin.PlatformViewCreationOptions, targetIdentity?: OpenFin.Identity) => Promise<import("/root/work/node_modules/openfin-adapter/src/api/view/Instance").View>' is not as signable to type '(view: string | OpenFin.PlatformViewCreationOptions, targetIdentity?: OpenFin.Identity) => Promise<import("/root/work/how-to/customize-workspace/node_modules/openfin-adapter/src/api/view/Instance").View>'.
```
```
ERROR in /root/work/how-to/customize-workspace/client/src/platform.ts
./client/src/platform.ts 17:8-31
[tsl] ERROR in /root/work/how-to/customize-workspace/client/src/platform.ts(17,9)
      TS2322: Type '(InteropBroker: Constructor<InteropBroker>, provider: any, options: any, ...args: any[]) => Promise<PlatformInteropBroker>' is not assignable to type 'OverrideCallback<InteropBroker, InteropBroker>'.
```
```
ERROR in /root/work/how-to/customize-workspace/client/src/bootstrapper.ts
45:8-18
[tsl] ERROR in /root/work/how-to/customize-workspace/client/src/bootstrapper.ts(45,9)
      TS2322: Type '(view: string | OpenFin.PlatformViewCreationOptions, targetIdentity?: OpenFin.Identity) => Promise<import("/root/work/node_modules/openfin-adapter/src/api/view/Instance").View>' is not as signable to type '(view: string | OpenFin.PlatformViewCreationOptions, targetIdentity?: OpenFin.Identity) => Promise<import("/root/work/how-to/customize-workspace/node_modules/openfin-adapter/src/api/view/Instan
ce").View>'.
```
```
ERROR in /root/work/how-to/customize-workspace/client/src/platform.ts
17:8-31
[tsl] ERROR in /root/work/how-to/customize-workspace/client/src/platform.ts(17,9)
      TS2322: Type '(InteropBroker: Constructor<InteropBroker>, provider: any, options: any, ...args: any[]) => Promise<PlatformInteropBroker>' is not assignable to type 'OverrideCallback<InteropBroker, InteropBroker>'.
```

As for build times:
- `node:14` with `npm v6.x` without workspaces:
  ```
  $ time npm install
  npm notice created a lockfile as package-lock.json. You should commit this file.
  npm WARN openfin-workspace-starter@7.0.0 No repository field.

  added 35 packages from 16 contributors and audited 36 packages in 1.977s

  4 packages are looking for funding
    run `npm fund` for details

  found 0 vulnerabilities


  real    0m2.434s
  user    0m1.850s
  sys     0m0.180s
  ```
  ```
  $ time npm run build
  [...]

  real    1m46.074s                           
  user    2m0.135s
  sys     0m11.802s
  ```
  For a total of **1m48.508s**.
- `node:16` with `npm v8.x` with workspaces:
  ```
  $ time npm install

  added 306 packages, and audited 316 packages in 2m

  38 packages are looking for funding
    run `npm fund` for details

  found 0 vulnerabilities
  npm notice·
  npm notice New minor version of npm available! 8.5.5 -> 8.10.0
  npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.10.0
  npm notice Run npm install -g npm@8.10.0 to update!
  npm notice·

  real    1m30.235s
  user    0m22.123s
  sys     0m5.059s
  ```
  ```
  $ time npm run build-client --workspaces

  real    0m41.971s
  user    1m12.183s
  sys     0m2.045s
  ```
  For a total of **2m12.206s**.

All in all, while workspaces do seem to be a cleaner solution also do take a little longer and will require some extra work in fixing the HOWTO package errors.

Please advise how you'd like to proceed.

Reply:
Interestingly I just tried the same tests and got the opposite result, the workspaces version was almost a minute quicker.

You can also see a Workspaces GHA https://github.com/built-on-openfin/workspace-starter/actions/runs/2384359479 was 2m15s
whereas Legacy GHA https://github.com/built-on-openfin/workspace-starter/actions/runs/2384287219 was 3m22s

**Local Build Legacy**

time npm install

real    0m6.575s
user    0m0.885s
sys     0m0.357s

time npm run package-how-tos
 
real    2m51.184s
user    1m13.711s
sys     0m58.979s

Total = 2m57.759s


**Local Build Workspaces**

time npm install

real    0m18.225s
user    0m9.255s
sys     0m10.023s

time npm run build-client --workspaces

real    1m40.779s
user    0m48.966s
sys     0m12.835s

time npm run package-how-tos

real    0m4.164s
user    0m0.311s
sys     0m0.750s

Total = 2m3.168s

I have also updated the https://github.com/built-on-openfin/workspace-starter/tree/dev/martyn/package-using-workspaces branch to align some of the package versions so it builds successfully.

Reply:
I suppose the turnaround time is heavily dependent on the network (fetching dependencies being the heaviest job) and I am on a GSM network connection which does not offer a very consistent speed. So my results may have been skewed.

In any case, I have no problem adding the workspaces solution. The only thing that I see as a blocker are those errors I pasted above.